pkg/trait/toleration.go

```diff
@@ -0,0 +1,121 @@
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to You under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trait
+
+import (
+	"fmt"
+	"strconv"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	v1 "github.com/apache/camel-k/pkg/apis/camel/v1"
+	"github.com/apache/camel-k/pkg/util"
+)
+
+// This trait will enable a Toleration.
+// Tolerations are applied to pods, and allow (but do not require) the pods to schedule onto nodes with matching taints.
+// See https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/ for more details.
+//
+// It's disabled by default.
+//
+// +camel-k:trait=toleration
+type tolerationTrait struct {
+	BaseTrait `property:",squash"`
+	// The key to match the Taint
+	Key string `property:"key" json:"key,omitempty"`
+	// The operator (Equal | Exists)
+	Operator string `property:"operator" json:"operator,omitempty"`
+	// The value to match if Equal operator selected
+	Value string `property:"value" json:"value,omitempty"`
+	// The effect that will be set on the Pod (NoExecute | NoSchedule | PreferNoSchedule)
+	Effect string `property:"effect" json:"effect,omitempty"`
+	// How long that Pod stays bound to a failing or unresponsive Node
+	TolerationSeconds string `property:"seconds" json:"seconds,omitempty"`
+}
+
+func newTolerationTrait() Trait {
+	return &tolerationTrait{
+		BaseTrait: NewBaseTrait("toleration", 1200),
+	}
+}
+
+func (t *tolerationTrait) Configure(e *Environment) (bool, error) {
+	if util.IsNilOrFalse(t.Enabled) {
+		return false, nil
+	}
+
+	if t.Key == "" {
+		return false, fmt.Errorf("missing key for toleration trait")
+	}
+
+	if t.Operator != "Equal" && t.Operator != "Exists" {
+		return false, fmt.Errorf("expected Equal or Exists for operator toleration trait, got %v", t.Operator)
+	}
+
+	if t.Operator == "Equal" && t.Value == "" {
+		return false, fmt.Errorf("missing value for equal operator toleration trait")
+	}
+
+	if t.Effect != "NoExecute" && t.Effect != "NoSchedule" && t.Effect != "PreferNoSchedule" {
+		return false, fmt.Errorf("expected NoExecute, NoSchedule or PreferNoSchedule for effect toleration trait, got %v", t.Effect)
+	}
+
+	return e.IntegrationInPhase(v1.IntegrationPhaseDeploying, v1.IntegrationPhaseRunning), nil
+}
+
+func (t *tolerationTrait) Apply(e *Environment) (err error) {
+	var deployment *appsv1.Deployment
+	e.Resources.VisitDeployment(func(d *appsv1.Deployment) {
```

Comment:
What about the other deployment strategies, like Knative Service or CronJob?

Reply:
Honestly I am not familiar with that. I mimicked the logic from `Affinity` trait.

Reply:
Right, I think the Affinity trait pre-dates the CronJob strategy, and possibly miss to integrate with the KnativeService one.

From a functional standpoint, I think tolerations work for both the KnativeService and the CronJob strategies. So technically, it is a matter of visiting also the KnativeService and CronJob resources, and add the tolerations to their PodTemplateSpec. You can find an example in the Container trait.

That should probably be done for the Affinity trait as well. We may find a better mechanism to avoid that visiting thing being spread in multiple places.